### ISSUE TITLE
Automated Deployment Pipeline

### ISSUE THREAD

**Issue by aaleksandraristic** [IanTDuncan/MealTime#38]
Describe how your continuous integration and automated deployment pipeline works. If there is any piece of the process that can't be automated please explain why.

**Comment by aaleksandraristic** [comment]
Here is one example with good graph, we can check it out:
https://www.cyberark.com/what-is/ci-cd-pipeline/

**Comment by BeepDroid** [comment]
Posting some notes here for easier understanding and reference while building the pipeline (links included) 

"When planning your deployment pipeline, there are three essential components to include:

 - **Build Automation (Continuous Integration)**: Build automation, also referred to as [Continuous Integration] or CI for short, are automated steps within development designed for continuous integration – the compilation, building, and merging of new code.

- **Test Automation:** Test automation relies on the creation of custom-written tests that are automatically triggered throughout a deployment pipeline and work to verify new compiled code against your organization’s predetermined acceptance criteria.

-**Deploy Automation (Continuous Deployment/Delivery):** Like continuous integration, deploy automation with [Continuous Deployment] (CD for short) helps expedite code delivery by automating the process of releasing code to a shared repository, and then automatically deploying the updates to a development or production environment."
(https://www.pagerduty.com/resources/learn/what-is-continuous-integration/)

Some expansion on CI and CD:
"**Continuous integration** is the developer practice of frequently (at worst, daily) integrating all code changes into a single, authoritative code branch, from which [production environments]are deployed/production software is built. This practice enables early problem detection and provides immediate feedback on code quality, allowing for issues to be identified, diagnosed, and fixed immediately, without interruption to a production environment. " (https://www.pagerduty.com/resources/learn/what-is-continuous-integration/)

"Your customers are waiting. They’re eagerly anticipating much-needed changes to your application, and you don’t want them to wait any longer. Enter **Continuous Deployment.** [Continuous Deployment (CD)] is a strategy whereby software testing and sendoff to consumers is automated. It’s a way to make the [CI/CD pipeline] faster, more efficient, and more agile — and give customers what they want, when they want it.

In essence, Continuous Deployment is the automation of the deployment process, whereby changes to software code are tested and deployed automatically." (https://www.pagerduty.com/resources/learn/continuous-deployment/)

CD(Continuous Deployment) vs Continues Delivery:
"They both refer to a level of software development automation. Where they differ, typically, is where the automation ends. In the Continuous Delivery stage, code changes are automated, but the actual deployment is manually completed; in Continuous Deployment, the entire deployment process is automated, from code review to deployment to a live environment. Once the new code is tested and validated, it’s sent out to a live environment." (https://www.pagerduty.com/resources/learn/continuous-deployment/)

CD vs CI:
"The entire process of creating/changing code and releasing it into the world is called the CI/CD Pipeline, because this describes the steps changes to the software application must follow in order to reach consumers.

In a nutshell, Continuous Integration ensures that code changes made from different team members, places and time zones all manage to place nicely together. Imagine you’re baking a cake with a variety of other people. Each of you is adding ingredients you know belong, but none of you are aware of the others have already added those ingredients. Continuous Integration keeps track of changes being made and verifies to make sure none of them conflict with one another. It also ensures that all changes are saved and merged in a central repository — a recipe, if you will." (https://www.pagerduty.com/resources/learn/continuous-deployment/)

**Comment by BeepDroid** [comment]
https://youtu.be/z_oLA56gE7o?si=TFjSSu9Vn85T_6mE 

--Really great video for reference.

**Comment by BeepDroid** [comment]
[Agile development.pdf](https://github.com/IanTDuncan/MealTime/files/13112456/Agile.development.pdf)
Here is the PDF for the Pipeline, ignore the name of the file, it's just the program I used to illustrate the pipeline.

**Comment by BeepDroid** [comment]
![image](https://github.com/IanTDuncan/MealTime/assets/71675347/e579627e-144c-4af5-bd90-70a1ccbf2414)
And, as always, screenshot for convienience!

**Comment by Cade5480** [comment]
This all looks really good!
